File: client-ios/SihlClient/Libre/Services/LibreStorageService.swift
import Foundation

/// Lokale Persistenz für Libre-Daten (Offline-First, kein Backend).
/// Token → Keychain (sicher), Readings → JSON-Datei im Documents-Verzeichnis.
struct LibreStorageService {
    private enum Keys {
        static let token     = "libre_token"       // Keychain
        static let expires   = "libre_expires"     // UserDefaults (ms)
        static let email     = "libre_email"       // UserDefaults
        static let patientId = "libre_patient_id"  // UserDefaults
        static let region    = "libre_region"      // UserDefaults
        static let lastSync  = "libre_last_sync"   // UserDefaults
    }

    private let readingsFile: URL = {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        return docs.appendingPathComponent("libre_readings.json")
    }()

    // MARK: - Auth

    func saveAuth(email: String, ticket: LibreAuthTicket,
                  patientId: String, region: LibreRegion) {
        KeychainStore.set(ticket.token, for: Keys.token)
        let ud = UserDefaults.standard
        ud.set(Int(ticket.expires.timeIntervalSince1970 * 1000), forKey: Keys.expires)
        ud.set(email, forKey: Keys.email)
        ud.set(patientId, forKey: Keys.patientId)
        ud.set(region.rawValue, forKey: Keys.region)
    }

    func loadAuthTicket() -> LibreAuthTicket? {
        guard let token = KeychainStore.get(Keys.token),
              !token.isEmpty else { return nil }
        let ms = UserDefaults.standard.integer(forKey: Keys.expires)
        let expires = ms > 0
            ? Date(timeIntervalSince1970: Double(ms) / 1000)
            : Date().addingTimeInterval(60 * 60 * 24 * 180)
        return LibreAuthTicket(token: token, expires: expires)
    }

    func loadEmail() -> String?     { UserDefaults.standard.string(forKey: Keys.email) }
    func loadPatientId() -> String? { UserDefaults.standard.string(forKey: Keys.patientId) }

    func loadRegion() -> LibreRegion {
        guard let raw = UserDefaults.standard.string(forKey: Keys.region),
              let region = LibreRegion(rawValue: raw) else { return .eu }
        return region
    }

    func clearAuth() {
        KeychainStore.remove(Keys.token)
        [Keys.expires, Keys.email, Keys.patientId, Keys.region]
            .forEach { UserDefaults.standard.removeObject(forKey: $0) }
    }

    // MARK: - Readings (JSON-Datei im Documents-Verzeichnis)

    func saveReadings(_ readings: [GlucoseReading]) {
        guard let data = try? JSONEncoder().encode(readings) else { return }
        try? data.write(to: readingsFile, options: .atomic)
        UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: Keys.lastSync)
    }

    func loadReadings() -> [GlucoseReading] {
        guard let data = try? Data(contentsOf: readingsFile),
              let readings = try? JSONDecoder().decode([GlucoseReading].self, from: data)
        else { return [] }
        return readings
    }

    func loadLastSync() -> Date? {
        let ts = UserDefaults.standard.double(forKey: Keys.lastSync)
        return ts > 0 ? Date(timeIntervalSince1970: ts) : nil
    }

    func clearReadings() {
        try? FileManager.default.removeItem(at: readingsFile)
        UserDefaults.standard.removeObject(forKey: Keys.lastSync)
    }
}

File: client-flutter/lib/services/libre_storage_service.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/glucose_reading.dart';
import '../models/libre_auth.dart';

/// Lokale Persistenz für Libre-Daten via SharedPreferences.
/// Alle Messdaten bleiben offline verfügbar; kein Backend nötig.
class LibreStorageService {
  static const _keyToken = 'libre_auth_token';
  static const _keyTokenExpires = 'libre_auth_expires';
  static const _keyEmail = 'libre_email';
  static const _keyRegion = 'libre_region';
  static const _keyReadings = 'libre_glucose_readings';
  static const _keyPatientId = 'libre_patient_id';
  static const _keyLastSync = 'libre_last_sync';

  // MARK: - Auth

  Future<void> saveAuth(String email, LibreAuthTicket ticket,
      String patientId, LibreRegion region) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_keyEmail, email);
    await prefs.setString(_keyToken, ticket.token);
    await prefs.setInt(
        _keyTokenExpires, ticket.expires.millisecondsSinceEpoch);
    await prefs.setString(_keyPatientId, patientId);
    await prefs.setString(_keyRegion, region.name);
  }

  Future<LibreAuthTicket?> loadAuthTicket() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString(_keyToken);
    final expiresMs = prefs.getInt(_keyTokenExpires);
    if (token == null || expiresMs == null) return null;
    return LibreAuthTicket(
      token: token,
      expires: DateTime.fromMillisecondsSinceEpoch(expiresMs),
    );
  }

  Future<String?> loadEmail() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_keyEmail);
  }

  Future<String?> loadPatientId() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_keyPatientId);
  }

  Future<LibreRegion> loadRegion() async {
    final prefs = await SharedPreferences.getInstance();
    final name = prefs.getString(_keyRegion) ?? 'eu';
    return LibreRegion.values.firstWhere((r) => r.name == name,
        orElse: () => LibreRegion.eu);
  }

  Future<void> clearAuth() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_keyToken);
    await prefs.remove(_keyTokenExpires);
    await prefs.remove(_keyEmail);
    await prefs.remove(_keyPatientId);
    await prefs.remove(_keyRegion);
  }

  // MARK: - Glukose-Readings

  Future<void> saveReadings(List<GlucoseReading> readings) async {
    final prefs = await SharedPreferences.getInstance();
    final json = readings.map((r) => jsonEncode(r.toJson())).toList();
    await prefs.setStringList(_keyReadings, json);
    await prefs.setString(_keyLastSync, DateTime.now().toIso8601String());
  }

  Future<List<GlucoseReading>> loadReadings() async {
    final prefs = await SharedPreferences.getInstance();
    final jsonList = prefs.getStringList(_keyReadings) ?? [];
    return jsonList.map((s) {
      try {
        return GlucoseReading.fromJson(jsonDecode(s) as Map<String, dynamic>);
      } catch (_) {
        return null;
      }
    }).whereType<GlucoseReading>().toList();
  }

  Future<DateTime?> loadLastSync() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_keyLastSync);
    return raw != null ? DateTime.tryParse(raw) : null;
  }

  Future<void> clearReadings() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_keyReadings);
    await prefs.remove(_keyLastSync);
  }
}
